### ISSUE TITLE
shared link lost after move the shared folder

### ISSUE THREAD

**Issue by dttpy** [owncloud/core#13776]
### Steps to reproduce
1. log in as admin
2. create a folder "pic1"
3. upload some files in "pic1"
4. share "pic1" with "tester1"
5. share the link of this folder
6. log in as tester1
7. create a folder "pics"
8. share folder "pics" with admin 
9. move shared folder "pic1" into pics
10. log in as admin again
11. move folder "pic1" manuelly into folder "pics"
12. go into folder "pics"

**Edit** (by @morrisjobke) I swapped 8 and 9 to correct this

### Expected behaviour
step 10: there is a folder "pic1" in the shared "pics" folder ???
step 12: folder "pic1" is still shared by link

### Actual behaviour
step 10: there is nothing in "pics" folder 
(maybe it is the logic because the ower of "pic1" is admin, and there is no need to have a duplicate one, but it is still a bit weird/different)

step 12: folder "pics1" is still shared with tester1, but it is no more shared by link 

### Server configuration
**Operating system:** Ubuntu 13.10

**Web server:** apache 2.4.6 (Ubuntu)

**Database:** SQLite

**PHP version:** 5.5.3-1ubuntu2.3

**ownCloud version:** 8.0.0-beta2

**Updated from an older ownCloud or fresh install:** owncloud 7.0.4

**List of activated apps:** standard apps of owncloud

**The content of config/config.php:**

```
<?php
define('DEBUG',true);
$CONFIG = array (
  'instanceid' => 'oc6d9cda6e26',

  'datadirectory' => '/var/www/owncloud8-beta2/data',
  'dbtype' => 'sqlite3',
  'version' => '8.0.0.4',
  'installed' => true,
  'theme' => '',
  'maintenance' => false,
  'maxZipInputSize' => 838860800,
  'allowZipDownload' => true,
  'loglevel' => '0',
  'trusted_domains' => 
  array (
    0 => '192.168.1.80',
  ),
  'appcodechecker' => false,

  );



```

**Are you using external storage, if yes which one:** no
**Are you using encryption:** yes

### Client configuration
**Browser:** firefox 35.0.1

**Operating system:** ubuntu 13.10

**Comment by MorrisJobke** [comment]
@dttpy I get following error in step 9 `Sharing "pics" failed, because it contains files shared with you!` Did you have any errors in the owncloud.log or javascript console in your browser?

**Comment by MorrisJobke** [comment]
So I can't reproduce anything here. Also it shouldn't be allowed to share something with the owner, because this would create a recursion.

**Comment by PVince81** [comment]
Swap steps 8 and 9

**Comment by MorrisJobke** [comment]
That's how it looks like after step9:

![bildschirmfoto von 2015-01-29 19 28 10](https://cloud.githubusercontent.com/assets/245432/5963488/0a9803de-a7ed-11e4-8e6c-c6bfcc034e87.png)

**Comment by MorrisJobke** [comment]
@PVince81 Nice one ... I will try.

**Comment by MorrisJobke** [comment]
Yes. Reproducible.

But I have no idea how this should be handled. What should the user expect if this happens. It would be even more weird if his shared folder disappears from the original location and appears at a completely new location o.O But forbidding the move would also be bad, because it restricts the user quite heavily. But maybe this is the only way to prohibit such stuff, but it is really hard to track if there is a bigger dependency cycle of shares: a shared to b shared to c shared to a -.-

> Swap steps 8 and 9

I fixed this in the message above

**Comment by PVince81** [comment]
I think it should behave the same as when 8 and 9 are swapped.
Since the user is not allowed to share the folder because it contains already received shares, then moving a received share folder should not be moveable into an already shared folder.

Basically, show the same message in that situation.

@schiesbn

**Comment by MorrisJobke** [comment]
@PVince81 The problem here is to detect this in case of resharing and stuff like that. I guess this is solvable once me migrated to the "simple sharing" logic (whit adding reshares to the original file).

**Comment by PVince81** [comment]
Side note: the reason why the link share is lost is most likely due to file id stability issues when moving stuff into a shared folder. But if we say doing this kind of move shouldn't be allowed in the first place, then this wouldn't be an issue.

**Comment by PVince81** [comment]
Ref #13312 (stable file id issues)

**Comment by PVince81** [comment]
Something to verify in 8.1 RC2

**Comment by PVince81** [comment]
Still happens on 8.1 RC2. I actually mixed it up with another issue, that's why I thought it was fixed.

**Comment by PVince81** [comment]
Moving a shared folder into another shared folder should be forbidden, so a warning needs to be shown instead of carrying on. @schiesbn @icewind1991

**Comment by PVince81** [comment]
@rullzer

**Comment by PVince81** [comment]
Step 11 "move folder "pic1" manuelly into folder "pics" doesn't work any more on master / 8.2.
It says 'Could not move "pic1" - File with this name already exists'
and webdav says "forbidden".

At least the original bug situation cannot be reproduced.

**Comment by PVince81** [comment]
Hmm I think the real bug here is that one should not be able to move a received mount point into an outgoing share.
Basically step 9 should be forbidden.

**Comment by PVince81** [comment]
I thought it was a regression, but it seems it also didn't prevent this back in stable7 or 7.0.4.

**Comment by PVince81** [comment]
`isTargetAllowed` here https://github.com/owncloud/core/blob/v8.1.3/lib/private/files/view.php#L1599 only checks whether the target folder itself is a mount point.
In the test case at step9 "pics" is not a mount point, just a local folder that happens to be shared.
(if you first move "pic1" into "pics" and try sharing it afterwards, you'll get an expected error)

**Comment by PVince81** [comment]
Fix is here https://github.com/owncloud/core/pull/19525
